README
NAME
    Alien::FLTK - Build and use the stable 1.3.x branch of the Fast Light
    Toolkit

Description
    This distribution builds and installs libraries for the (stable) `1.3.x'
    branch of the FLTK GUI toolkit.

Synopsis
        use Alien::FLTK;
        use ExtUtils::CBuilder;
        my $AF  = Alien::FLTK->new();
        my $CC  = ExtUtils::CBuilder->new();
        my $SRC = 'hello_world.cxx';
        open(my $FH, '>', $SRC) || die '...';
        syswrite($FH, <<'') || die '...'; close $FH;
          #include <FL/Fl.H>
          #include <FL/Fl_Window.H>
          #include <FL/Fl_Box.H>
          int main(int argc, char **argv) {
            Fl_Window *window = new Fl_Window(300,180);
            Fl_Box *box = new Fl_Box(FL_UP_BOX, 20, 40, 260, 100, "Hello, World!");
            box->labelfont(FL_BOLD + FL_ITALIC);
            box->labelsize(36);
            box->labeltype(FL_SHADOW_LABEL);
            window->end();
            window->show(argc, argv);
            return Fl::run();
        }

        my $OBJ = $CC->compile('C++'                => 1,
                               source               => $SRC,
                               include_dirs         => [$AF->include_dirs()],
                               extra_compiler_flags => $AF->cxxflags()
        );
        my $EXE =
            $CC->link_executable(objects            => $OBJ,
                                 extra_linker_flags => $AF->ldflags());
        print system('./' . $EXE) ? 'Aww...' : 'Yay!';
        END { unlink grep defined, $SRC, $OBJ, $EXE; }

Constructor
    There are no per-object configuration options as of this version, but
    there may be in the future, so any new code using Alien::FLTK should
    create objects with the `new' constructor.

        my $AF = Alien::FLTK->new( );

Methods
    After creating a new Alien::FLTK object, use the following methods to
    gather information:

  `include_dirs'
        my @include_dirs = $AF->include_dirs( );

    Returns a list of the locations of the headers installed during the
    build process and those required for compilation.

  `library_path'
        my $lib_path = $AF->library_path( );

    Returns the location of the private libraries we made and installed
    during the build process.

  `cflags'
        my $cflags = $AF->cflags( );

    Returns additional C compiler flags to be used.

  `cxxflags'
        my $cxxflags = $AF->cxxflags( );

    Returns additional flags to be used to when compiling C++ using FLTK.

  `ldflags'
        my $ldflags = $AF->ldflags( qw[gl images] );

    Returns additional linker flags to be used. This method can
    automatically add appropriate flags based on how you plan on linking to
    fltk. Acceptable arguments are:

    `static'
        Returns flags to link against a static FLTK library.

        FLTK's license allows static linking, but Alien::FLTK does not build
        static libs. ...yet.

    `gl'
        Include flags to use GL.

        *This is an experimental option. Depending on your system, this may
        also include OpenGL or MesaGL.*

    `images'
        Include flags to use extra image formats (PNG, JPEG).

  `branch'
        my $revision = $AF->branch( );

    Returns the SVN branch of the source Alien::FLTK was built with. For the
    `2.0.x' branch of fltk, see Alien::FLTK.

  `revision'
        my $revision = $AF->revision( );

    Returns the SVN revision number of the source Alien::FLTK was built
    with.

  `capabilities'
        my $caps = $AF->capabilities( );

    Returns a list of capabilities supported by your Alien::FLTK
    installation. This list can be handed directly to `ldflags( )'.

  `config'
        my $configuration = $AF->config( );

    Returns a hashref containing the raw configuration data collected during
    build. This would be helpful when reporting bugs, etc.

Notes
  Requirements
    Prerequisites differ by system...

    Win32
        The fltk libs and Alien::FLTK both build right out of the box with
        MinGW. Further testing is needed for other setups.

    X11/*nix
        X11-based systems require several development packages. On Debian,
        these may be installed with:

            > sudo apt-get install libx11-dev
            > sudo apt-get install libxi-dev

        Additionally, the optional XCurser lib may be installed with:

            > sudo apt-get install libxcursor-dev

    Darwin/OSX
        Uh, yeah, I have no idea.

  Installation
    The distribution is based on Module::Build, so use the following
    procedure:

      > perl Build.PL
      > ./Build
      > ./Build test
      > ./Build install

  Support Links
    * Issue Tracker
        http://github.com/sanko/alien-fltk/issues

        Please only report Alien::FLTK related bugs to this tracker. For
        FLTK issues, use http://github.com/sanko/fltk-perl/issues/

    * Commit Log
        http://github.com/sanko/alien-fltk/commits/master

    * Homepage:
        http://sanko.github.com/fltk-perl/ is the homepage of the FLTK
        project.

    * License:
        http://www.perlfoundation.org/artistic_license_2_0

        See the License and Legal section of this document.

    * Mailing List
        Once I find someone to host a list for the FLTK project, I'll use it
        for Alien::FLTK too.

    * Repository
        http://github.com/sanko/alien-fltk/ and you are invited to fork it.

  Examples
    Please see the Synopsis and the files in the `/examples/'.

  Bugs
    Numerous, I'm sure.

  To Do
    Please see Alien::FLTK::Todo

See Also
    FLTK, Alien::FLTK2

Acknowledgments
    The FLTK Team - http://www.fltk.org/

Author
    Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

    CPAN ID: SANKO

License and Legal
    Copyright (C) 2009-2012 by Sanko Robinson <sanko@cpan.org>

    This program is free software; you can redistribute it and/or modify it
    under the terms of The Artistic License 2.0. See the LICENSE file
    included with this distribution or
    http://www.perlfoundation.org/artistic_license_2_0. For clarification,
    see http://www.perlfoundation.org/artistic_2_0_notes.

    When separated from the distribution, all POD documentation is covered
    by the Creative Commons Attribution-Share Alike 3.0 License. See
    http://creativecommons.org/licenses/by-sa/3.0/us/legalcode. For
    clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

    Alien::FLTK is based in part on the work of the FLTK project. See
    http://www.fltk.org/.



Changes
Revision history for Perl extension Alien-FLTK

{{$NEXT}}

v1.3.5 2020-01-21T03:15:23Z

	- Correct JSON::Tiny prereq in cpanfile

v1.3.4 2020-01-20T15:05:40Z

	- Pull fltk 1.3.x snapshots from github

v1.3.3 2016-04-18T15:27:53Z

	- Configure 64bit lib for relocation
	- New opengl example

v1.3.3 2016-04-14T01:45:26Z

    - Build with stable 1.3.3 release of FLTK
    - Resolve missing prereq (I hope)

Version 0.09635001 | 2012-07-22 08:40:13Z (Sun, 22 Jul 2012) | 5dae2c1

    Resolved Issues/Bug Fixes:
     * Define FL_LIBRARY during build
     * Work around Win32 bug

    Notes:
     * Builds against r9635 of the 1.3.x branch of fltk

Version 0.07704001 | 2010-09-18 01:55:22Z (Sat, 18 Sep 2010) | 70e72ea

    Resolved Issues/Bug Fixes:
     * #define symbols for dirent location were malformed

Version 0.07694001 | 2010-09-04 03:56:05Z (Sat, 04 Sep 2010) | 45bb2d7

    Selected Upstream Changes:
     * [r7704] Fixed Windows class name (xclass) setup and caching mechanism.
       This was not yet completely converted to UTF-8. Cleaned up related
       code, removed old comments.
       Tested and confirmed that the xclass() setup works and that different
       icons can be set for different window classes (as it was in FLTK
       1.1.10).
     * [r7697] Full support of UTF-8 text for PostScript printing/output in 2
         ways:
          - Latin characters (unicode <= 17F) and a few other symbols, among
            which the ?\226?\130?\172 (Euro), are output using vectorial fonts
          - All other characters are output as a bitmap copy of their
            offscreen rendition.
     * [r7696] Simpler PostScript output of Fl_Bitmap objects
     * [r7695] PostScript output of Fl_Pixmap objects: accept now arbitrarily
       large pixmaps

Version 0.07694001 | 2010-09-04 03:56:05Z (Sat, 04 Sep 2010) | 45bb2d7

    Notes:
     * Builds against r7694 of the 1.3.x branch of fltk

Version 0.07063001 | 2010-02-15 22:06:30Z (Mon, 15 Feb 2010) | b0766e6

    Resolved Issues/Bug Fixes:
     * Untested CygWin configuration and build fixes

    Notes:
     * For now, we only support Win32-ish CygWin builds. In the future, we may
       include X11-like builds. But don't count on it.

Version 0.07063 | 2010-02-13 16:54:31Z (Sat, 13 Feb 2010) | 0bfc53b

    Resolved Issues/Bug Fixes:
     * Rather than X11, OSX/Darwin uses Carbon for graphics so we no longer
       search for X11-related libs

    Documentation/Sample Code/Test Suite:
     * The XCursor lib is now noted as being optional in POD and README

    Notes:
     * This dist also includes other OS X 'fixes' which are completly untested
       and will probably fail
     * Upgraded fltk lib to svn r7063. See the following changelogs for more:
       - http://fltk.org/articles.php?L967 (r7020 snapshot)
       - http://fltk.org/articles.php?L968 (r7034 snapshot)
       - http://fltk.org/articles.php?L970 (r7037 snapshot)
       - http://fltk.org/articles.php?L973 (r7063 snapshot)

Version 0.07008017 | 2010-02-10 07:18:40Z (Wed, 10 Feb 2010) | 364f273

     Resolved Issues/Bug Fixes:
     * X11-related libs are now linked in the correct, working order (I hope)

Version 0.07008016 | 2010-02-09 21:35:54Z (Tue, 09 Feb 2010) | 37735bc

    Internal/Behavioral Changes:
     * Alien::FLTK has been split into two projects. See Notes below.
     * cflags() and cxxflags() no longer include list of include directories
     * Renamed include_path() => include_dirs() now that it returns all
          required include directories
     * Internal method config('blah') now returns blah's data in the config
          hash
     * Compile using Quartz drawing instead of X11 on Apple OS X
     * Four months of stuff I forgot to keep the changelog up to date for. You
           could check out the commit logs if you're bored.

    Notes:
     * When stable, this will be an API-breakin' release
     * Alien::FLTK has been split into two projects. The first, Alien::FLTK
           will now build the stable 1.3.x branch of FLTK while Alien::FLTK2
           will build the 2.0.x branch.

Version 0.06921 | 2009-10-11 05:50:42Z (Sun, 11 Oct 2009) | 335da88

    Internal/Behavioral Changes:
     * Upgrading to r6921
     * include_path() and library_path() methods are (hopefully) less flimsey
     * Slightly better platform handling

Version 0.06916 | 2009-10-11 05:50:42Z (Sun, 11 Oct 2009) | 335da88

    Resolved Issues/Bug Fixes:
     * lib/header checks (should) now work on Unix
     * configure action /should/ work on Solaris

    Internal/Behavioral Changes:
     * Upgrading to r6916
     * We configure fltk before creating Build script in case we run into
          trouble (typically missing libs on *nix) and need to exit(0) so test
          results in NA

    Documentation/Sample Code/Test Suite:
     * Full OO syntax is now used in docs, tests, and examples
     * Trimmed down t/0000_use/0002_exe.t so tests don't fail under tunneled X11

    Notes:
     * I'll get this right eventually... ((sigh))

Version 0.06879 | 2009-10-04 04:46:04Z (Sun, 04 Oct 2009) | b9cf221

    Resolved Issues/Bug Fixes:
     * None yet

    Internal/Behavioral Changes:
     * More spitballing with *nix

    Documentation/Sample Code/Test Suite:
     * capabilities() and config() methods are now documented

    Notes:
     * Upgrading to fltk2 r6858 which comes with a modified license which
         remains "in the spirit of the original FLTK 2 license, but at the
         same time [complies with] Debian license requirements."

Version 0.06844 | 2009-09-03 05:03:50Z (Thu, 03 Sep 2009) | 7ea0002

     Resolved Issues/Bug Fixes:
     * It actually works (...on Windows, at least)

    Internal/Behavioral Changes:
     * Alternatives are presented when we fail to download fltk's source
     * Building fltk2 using ExtUtils::CBuilder and manual configuration

    Documentation/Sample Code/Test Suite:
     * New GL example: example/0003_gl.pl

inc/MBTFLTK.pm
package MBTFLTK;
use strict;
use warnings;
use Exporter 5.57 'import';
our @EXPORT = qw/Build Build_PL/;
use CPAN::Meta;
use ExtUtils::Config 0.003;
use ExtUtils::Helpers 0.020
    qw/make_executable split_like_shell man1_pagename man3_pagename detildefy/;
use ExtUtils::Install qw/pm_to_blib install/;
use ExtUtils::InstallPaths 0.002;
use File::Basename qw/basename dirname/;
use File::Find ();
use File::Path qw/mkpath/;
use File::Spec::Functions qw/catfile catdir rel2abs abs2rel splitdir/;
use Getopt::Long qw/GetOptions/;
use JSON::PP 2 qw/encode_json decode_json/;
use HTTP::Tiny;
use Archive::Extract;
use File::pushd;
use File::Copy;
use File::Copy::Recursive qw[dircopy];
use Devel::CheckBin qw[can_run];

sub write_file {
    my ($filename, $mode, $content) = @_;
    open my $fh, ">:$mode", $filename or die "Could not open $filename: $!\n";
    print $fh $content;
}

sub read_file {
    my ($filename, $mode) = @_;
    open my $fh, "<:$mode", $filename or die "Could not open $filename: $!\n";
    return do { local $/; <$fh> };
}

sub get_meta {
    my ($metafile) = grep { -e $_ } qw/META.json META.yml/
        or die "No META information provided\n";
    return CPAN::Meta->load_file($metafile);
}

sub manify {
    my ($input_file, $output_file, $section, $opts) = @_;
    return if -e $output_file && -M $input_file <= -M $output_file;
    my $dirname = dirname($output_file);
    mkpath($dirname, $opts->{verbose}) if not -d $dirname;
    require Pod::Man;
    Pod::Man->new(section => $section)
        ->parse_from_file($input_file, $output_file);
    print "Manifying $output_file\n"
        if $opts->{verbose} && $opts->{verbose} > 0;
    return;
}

sub process_xs {
    my ($source, $options) = @_;
    die "Can't build xs files under --pureperl-only\n"
        if $options->{'pureperl-only'};
    my (undef, @dirnames) = splitdir(dirname($source));
    my $file_base = basename($source, '.xs');
    my $archdir = catdir(qw/blib arch auto/, @dirnames, $file_base);
    my $c_file = catfile('lib', @dirnames, "$file_base.c");
    require ExtUtils::ParseXS;
    ExtUtils::ParseXS::process_file(filename   => $source,
                                    prototypes => 0,
                                    output     => $c_file
    );
    my $version = $options->{meta}->version;
    require ExtUtils::CBuilder;
    my $builder
        = ExtUtils::CBuilder->new(config => $options->{config}->values_set);
    my $ob_file = $builder->compile(
          source  => $c_file,
          defines => {VERSION => qq/"$version"/, XS_VERSION => qq/"$version"/}
    );
    mkpath($archdir, $options->{verbose}, oct '755') unless -d $archdir;
    return
        $builder->link(
        objects => $ob_file,
        lib_file =>
            catfile($archdir, "$file_base." . $options->{config}->get('dlext')
            ),
        module_name => join '::',
        @dirnames,
        $file_base
        );
}

sub get_lib {
    my ($meta) = @_;
    my $location;
    my $index = 'http://fltk.org/pub/fltk/snapshots/';
    my $snaps = qr[fltk-1.3.x-r([\d\.]+)\.tar\.gz];
    {
        print "Finding most recent version...";
        my $response = HTTP::Tiny->new->get($index);
        if (0 && $response->{success}) {

            # Snapshots don't contain fltk-config script
            my ($version)
                = sort { $b <=> $a } ($response->{content} =~ /$snaps/g);
            $location = $index . 'fltk-1.3.x-r' . $version . '.tar.gz';
            print " r$version\n";
        }
        else {
            print " Hrm. Grabbing latest stable release\n";
            $location
                = 'http://fltk.org/pub/fltk/1.3.3/fltk-1.3.3-source.tar.gz';
        }
    }
    my $file = basename($location);
    {
        print "Downloading $file...";
        my $response = HTTP::Tiny->new->mirror($location, $file);
        if ($response->{success}) {
            print " Done\n";
            return $file;
        }
    }
    exit !!print " Fail!";
}

sub build_lib {
    my ($options) = @_;
    my (%libinfo, $dir);
    my $meta = $options->{meta};
    my $cwd  = rel2abs './';       # XXX - use Cwd;

    # This is an ugly cludge. A working, ugly cludge though. :\
    if (!-d 'share') {
        mkpath('share', $options->{verbose}, oct '755') unless -d 'share';
        $dir = tempd();
        $libinfo{archive} = get_lib($meta->custom('x_alien'));
        print "Extracting...";
        my $ae = Archive::Extract->new(archive => $libinfo{archive});
        exit print " Fail! " . $ae->error if !$ae->extract();
        print " Done\nConfigure...\n";
        chdir($ae->extract_path);
        system q[sh ./configure];
        $libinfo{cflags}     = `sh ./fltk-config --cflags --optim`;
        $libinfo{cxxflags}   = `sh ./fltk-config --cxxflags --optim`;
        $libinfo{ldflags}    = `sh ./fltk-config --ldflags`;
        $libinfo{ldflags_gl} = `sh ./fltk-config --ldflags --use-gl`;
        $libinfo{ldflags_gl_images}
            = `sh ./fltk-config --ldflags --use-gl --use-images`;
        $libinfo{ldflags_images} = `sh ./fltk-config --ldflags --use-images`;

        # XXX - The following block is a mess!!!
        chdir 'src';
        my $gmake = can_run('gmake');
        my $make  = can_run('make');
        printf "Checking for gmake... %s\n", ($gmake ? 'yes' : 'no');
        printf "Checking for make... %s\n",  ($make  ? 'yes' : 'no');
        #system(($gmake ? 'g' : '') . q[make -ns > build.sh]);
        #system q[sh build.sh];
        system(($gmake ? 'g' : '') . q[make -j 10]);
        chdir '..';
        my $archdir = catdir($cwd, qw[share]);
        mkpath($archdir, $options->{verbose}, oct '755') unless -d $archdir;

        # XXX - Copy FL  => shared dir
        dircopy rel2abs('FL'), catdir($archdir, 'include', 'FL')
            or die $!;
        copy
            rel2abs(catdir('config.h')),
            catdir($archdir, 'include', 'config.h')
            or die $!;
        dircopy rel2abs('lib'), catdir($archdir, 'lib')
            or die $!;
        #

        write_file(catfile($archdir, qw[config.json]),
                   'utf8', encode_json(\%libinfo));
    }
}

sub find {
    my ($pattern, $dir) = @_;
    my @ret;
    File::Find::find(sub { push @ret, $File::Find::name if /$pattern/ && -f },
                     $dir)
        if -d $dir;
    return @ret;
}
my %actions = (
    build => sub {
        my %opt = @_;
        system $^X, $_ and die "$_ returned $?\n" for find(qr/\.PL$/, 'lib');
        my %modules
            = map { $_ => catfile('blib', $_) } find(qr/\.p(?:m|od)$/, 'lib');
        my %scripts = map { $_ => catfile('blib', $_) } find(qr//, 'script');
        build_lib(\%opt);
        my %shared = map {
            $_ => catfile(qw/blib lib auto share dist/,
                          $opt{meta}->name, abs2rel($_, 'share'))
        } find(qr//, 'share');
        pm_to_blib({%modules, %scripts, %shared}, catdir(qw/blib lib auto/));
        make_executable($_) for values %scripts;
        mkpath(catdir(qw/blib arch/), $opt{verbose});
        process_xs($_, \%opt) for find(qr/.xs$/, 'lib');

        if (   $opt{install_paths}->install_destination('libdoc')
            && $opt{install_paths}->is_default_installable('libdoc'))
        {   manify($_,
                   catfile('blib', 'bindoc', man1_pagename($_)),
                   $opt{config}->get('man1ext'), \%opt)
                for keys %scripts;
            manify($_,
                   catfile('blib', 'libdoc', man3_pagename($_)),
                   $opt{config}->get('man3ext'), \%opt)
                for keys %modules;
        }
    },
    test => sub {
        my %opt = @_;
        die "Must run `./Build build` first\n" if not -d 'blib';
        require TAP::Harness;
        my $tester = TAP::Harness->new(
            {verbosity => $opt{verbose},
             lib       => [
                 map {
                     rel2abs(catdir(qw/blib/, $_))
                 } qw/arch lib/
             ],
             color => -t STDOUT
            }
        );
        $tester->runtests(sort +find(qr/\.t$/, 't'))->has_errors and exit 1;
    },
    install => sub {
        my %opt = @_;
        die "Must run `./Build build` first\n" if not -d 'blib';
        install($opt{install_paths}->install_map,
                @opt{qw/verbose dry_run uninst/});
    },
);

sub Build {
    my $action = @ARGV && $ARGV[0] =~ /\A\w+\z/ ? shift @ARGV : 'build';
    die "No such action '$action'\n" if not $actions{$action};
    unshift @ARGV, @{decode_json(read_file('_build_params', 'utf8'))};
    GetOptions(\my %opt,
               qw/install_base=s install_path=s% installdirs=s destdir=s prefix=s config=s% uninst:1 verbose:1 dry_run:1 pureperl-only:1 create_packlist=i/
    );
    $_ = detildefy($_)
        for grep {defined} @opt{qw/install_base destdir prefix/},
        values %{$opt{install_path}};
    @opt{'config', 'meta'}
        = (ExtUtils::Config->new($opt{config}), get_meta());
    $actions{$action}->(
          %opt,
          install_paths =>
              ExtUtils::InstallPaths->new(%opt, dist_name => $opt{meta}->name)
    );
}

sub Build_PL {
    my $meta = get_meta();
    printf "Creating new 'Build' script for '%s' version '%s'\n",
        $meta->name, $meta->version;
    my $dir = $meta->name eq 'MBTFLTK' ? '' : "use lib 'inc';";
    write_file('Build', 'raw',
               "#!perl\n$dir\nuse MBTFLTK;\n\$|++;\nBuild();\n");
    make_executable('Build');
    my @env
        = defined $ENV{PERL_MB_OPT} ?
        split_like_shell($ENV{PERL_MB_OPT})
        : ();
    write_file('_build_params', 'utf8', encode_json([@env, @ARGV]));
    $meta->save(@$_) for ['MYMETA.json'], ['MYMETA.yml' => {version => 1.4}];
}
1;

=head1 SEE ALSO

L<Module::Build::Tiny>

=head1 ORIGINAL AUTHORS

=over 4

=item *

Leon Timmermans <leont@cpan.org>

=item *

David Golden <dagolden@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Leon Timmermans, David Golden.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


examples/synopsis.pl
use lib '../../blib';
$|++;
use Alien::Fltk;
use ExtUtils::CBuilder;
my $AF  = Alien::Fltk->new();
my $CC  = ExtUtils::CBuilder->new();
my $SRC = 'hello_world.cxx';
open(my $FH, '>', $SRC) || die '...';
syswrite($FH, <<'END') || die '...'; close $FH;
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>
int main(int argc, char **argv) {
  Fl_Window *window = new Fl_Window(300,180);
  Fl_Box *box = new Fl_Box(FL_UP_BOX, 20, 40, 260, 100, "Hello, World!");
  box->labelfont(FL_BOLD + FL_ITALIC);
  box->labelsize(36);
  box->labeltype(FL_SHADOW_LABEL);
  window->end();
  window->show(argc, argv);
  return Fl::run();
}
END
my $OBJ = $CC->compile('C++'                => 1,
                       source               => $SRC,
                       include_dirs         => [$AF->include_dirs()],
                       extra_compiler_flags => $AF->cxxflags()
);
my $EXE =
    $CC->link_executable(
         objects            => $OBJ,
         extra_linker_flags => '-L' . $AF->library_path . ' ' . $AF->ldflags('gl')
    );
print system('./' . $EXE) ? 'Aww...' : 'Yay!';
END { unlink grep defined, $SRC, $OBJ, $EXE; }


examples/0000_synopsis.pl
use strict;
use warnings;
{
    use Alien::FLTK;
    use ExtUtils::CBuilder;
    my $AF  = Alien::FLTK->new();
    my $CC  = ExtUtils::CBuilder->new();
    my $SRC = 'hello_world.cxx';
    open(my $FH, '>', $SRC) || die '...';
    syswrite($FH, <<'') || die '...'; close $FH;
      #include <FL/Fl.H>
      #include <FL/Fl_Window.H>
      #include <FL/Fl_Box.H>
      int main(int argc, char **argv) {
        Fl_Window *window = new Fl_Window(300,180);
        Fl_Box *box = new Fl_Box(FL_UP_BOX, 20, 40, 260, 100, "Hello, World!");
        box->labelfont(FL_BOLD + FL_ITALIC);
        box->labelsize(36);
        box->labeltype(FL_SHADOW_LABEL);
        window->end();
        window->show(argc, argv);
        return Fl::run();
    }

    my $OBJ = $CC->compile('C++'                => 1,
                           source               => $SRC,
                           include_dirs         => [$AF->include_dirs()],
                           extra_compiler_flags => $AF->cxxflags()
    );
    my $EXE =
        $CC->link_executable(objects            => $OBJ,
                             extra_linker_flags => $AF->ldflags());
    print system('./' . $EXE) ? 'Aww...' : 'Yay!';
    END { unlink grep defined, $SRC, $OBJ, $EXE; }
}

=pod

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2009 by Sanko Robinson E<lt>sanko@cpan.orgE<gt>

This program is free software; you can redistribute it and/or modify it under
the terms of The Artistic License 2.0. See the F<LICENSE> file included with
this distribution or http://www.perlfoundation.org/artistic_license_2_0.  For
clarification, see http://www.perlfoundation.org/artistic_2_0_notes.

When separated from the distribution, all POD documentation is covered by the
Creative Commons Attribution-Share Alike 3.0 License. See
http://creativecommons.org/licenses/by-sa/3.0/us/legalcode.  For
clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

=for git $Id$

=cut


lib/Alien/FLTK.pm
package Alien::FLTK;
{ $Alien::FLTK::VERSION = 'v1.3.4'; }
use strict;
use warnings;
use File::ShareDir qw[dist_dir];
use File::Spec::Functions qw[catdir canonpath];
use JSON::Tiny qw[decode_json];
use File::Slurp qw[read_file];
use lib '../../blib/lib', '../blib/lib', 'blib/lib', 'lib';

sub new {
    my ( $class, $overrides ) = @_;    # XXX - overrides are unsupported
    return
        bless decode_json(
        read_file( canonpath( catdir( dist_dir('Alien-FLTK'), 'config.json' ) ) ) ), shift;
}

sub include_dirs {
    my ($self) = @_;
    return canonpath( catdir( dist_dir('Alien-FLTK'), 'include' ) );
}

sub library_path {
    my ($self) = @_;
    return canonpath( catdir( dist_dir('Alien-FLTK'), 'lib' ) );
}
sub cflags   { +shift->{cflags} }
sub cxxflags { +shift->{cxxflags} }

sub ldflags {
    my ( $self, @args ) = @_;
    return +shift->{ join '_', 'ldflags', sort @args };
}
1;

=pod

=head1 NAME

Alien::FLTK - Build and use the stable 1.3.x branch of the Fast Light Toolkit

=head1 Description

This distribution builds and installs libraries for the (stable) C<1.3.x>
branch of the FLTK GUI toolkit.

=head1 Synopsis

    use Alien::FLTK;
    use ExtUtils::CBuilder;
    my $AF  = Alien::FLTK->new();
    my $CC  = ExtUtils::CBuilder->new();
    my $SRC = 'hello_world.cxx';
    open(my $FH, '>', $SRC) || die '...';
    syswrite($FH, <<'') || die '...'; close $FH;
      #include <FL/Fl.H>
      #include <FL/Fl_Window.H>
      #include <FL/Fl_Box.H>
      int main(int argc, char **argv) {
        Fl_Window *window = new Fl_Window(300,180);
        Fl_Box *box = new Fl_Box(FL_UP_BOX, 20, 40, 260, 100, "Hello, World!");
        box->labelfont(FL_BOLD + FL_ITALIC);
        box->labelsize(36);
        box->labeltype(FL_SHADOW_LABEL);
        window->end();
        window->show(argc, argv);
        return Fl::run();
    }

    my $OBJ = $CC->compile('C++'                => 1,
                           source               => $SRC,
                           include_dirs         => [$AF->include_dirs()],
                           extra_compiler_flags => $AF->cxxflags()
    );
    my $EXE =
        $CC->link_executable(
         objects            => $OBJ,
         extra_linker_flags => '-L' . $AF->library_path . ' ' . $AF->ldflags()
        );
    print system('./' . $EXE) ? 'Aww...' : 'Yay!';
    END { unlink grep defined, $SRC, $OBJ, $EXE; }

=head1 Constructor

There are no per-object configuration options as of this version, but there
may be in the future, so any new code using L<Alien::FLTK|Alien::FLTK> should
create objects with the C<new> constructor.

    my $AF = Alien::FLTK->new( );

=head1 Methods

After creating a new L<Alien::FLTK|Alien::FLTK> object, use the following
methods to gather information:

=head2 C<include_dirs>

    my @include_dirs = $AF->include_dirs( );

Returns a list of the locations of the headers installed during the build
process and those required for compilation.

=head2 C<library_path>

    my $lib_path = $AF->library_path( );

Returns the location of the private libraries we made and installed
during the build process.

=head2 C<cflags>

    my $cflags = $AF->cflags( );

Returns additional C compiler flags to be used.

=head2 C<cxxflags>

    my $cxxflags = $AF->cxxflags( );

Returns additional flags to be used to when compiling C++ using FLTK.

=head2 C<ldflags>

    my $ldflags = $AF->ldflags( qw[gl images] );

Returns additional linker flags to be used. This method can automatically add
appropriate flags based on how you plan on linking to fltk. Acceptable
arguments are:

=over

=item C<gl>

Include flags to use GL.

I<This is an experimental option. Depending on your system, this may also
include OpenGL or MesaGL.>

=item C<images>

Include flags to use extra image formats (PNG, JPEG).

=back

=head1 Notes

=head2 Requirements

Prerequisites differ by system...

=over

=item Win32

The fltk libs and L<Alien::FLTK|Alien::FLTK> both build right out of the box
with MinGW. Further testing is needed for other setups.

=item X11/*nix

X11-based systems require several development packages. On Debian, these may
be installed with:

    > sudo apt-get install libx11-dev
    > sudo apt-get install libxi-dev

Additionally, the optional XCurser lib may be installed with:

    > sudo apt-get install libxcursor-dev

=item Darwin/OSX

Uh, yeah, I have no idea.

=back

=head2 Installation

The distribution is based on L<Module::Build|Module::Build>, so use the
following procedure:

  > perl Build.PL
  > ./Build
  > ./Build test
  > ./Build install

=head2 Support Links

=over

=item * Issue Tracker

http://github.com/sanko/alien-fltk/issues

Please only report L<Alien::FLTK|Alien::FLTK> related bugs to this tracker.
For L<FLTK|FLTK> issues, use http://github.com/sanko/fltk-perl/issues/

=item * Commit Log

http://github.com/sanko/alien-fltk/commits/master

=item * Homepage:

http://sanko.github.com/fltk-perl/ is the homepage of the L<FLTK|FLTK>
project.

=item * License:

http://www.perlfoundation.org/artistic_license_2_0

See the L<License and Legal|/"License and Legal"> section of this document.

=item * Mailing List

Once I find someone to host a list for the L<FLTK|FLTK> project, I'll use it
for L<Alien::FLTK|Alien::FLTK> too.

=item * Repository

http://github.com/sanko/alien-fltk/ and you are invited to fork it.

=back

=head2 Examples

Please see the L<Synopsis|Alien::FLTK/"Synopsis"> and the files in the
C</examples/>.

=head2 Bugs

Numerous, I'm sure.

=head2 To Do

Please see L<Alien::FLTK::Todo|Alien::FLTK::Todo>

=head1 See Also

L<FLTK|FLTK>, L<Alien::FLTK2|Alien::FLTK2>

=head1 Acknowledgments

=over

=item The FLTK Team - http://www.fltk.org/

=back

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2009-2020 by Sanko Robinson E<lt>sanko@cpan.orgE<gt>

This program is free software; you can redistribute it and/or modify it under
the terms of The Artistic License 2.0. See the F<LICENSE> file included with
this distribution or http://www.perlfoundation.org/artistic_license_2_0.  For
clarification, see http://www.perlfoundation.org/artistic_2_0_notes.

When separated from the distribution, all POD documentation is covered by the
Creative Commons Attribution-Share Alike 3.0 License. See
http://creativecommons.org/licenses/by-sa/3.0/us/legalcode.  For
clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

L<Alien::FLTK|Alien::FLTK> is based in part on the work of the FLTK project.
See http://www.fltk.org/.

=cut
